File: mizhara-web/src/lib/pagination.ts
export type PaginationMeta = {
  page: number;
  limit: number;
  total: number;
  totalPages: number;
};

export function parsePagination(searchParams: URLSearchParams) {
  const page = Math.max(1, Number(searchParams.get("page")) || 1);
  let limit = Number(searchParams.get("limit")) || 10;
  if (limit > 50) limit = 50;
  const search = searchParams.get("search") || "";
  return { page, limit, skip: (page - 1) * limit, search };
}

export function buildPaginationMeta(page: number, limit: number, total: number): PaginationMeta {
  return { page, limit, total, totalPages: limit > 0 ? Math.ceil(total / limit) : 0 };
}

File: mizhara-backend/utils/pagination.go
package utils

import (
	"strconv"

	"mizhara-backend/constants"
)

type PaginationParams struct {
	Page   int
	Limit  int
	Skip   int
	Search string
}

type PaginationMeta struct {
	Page       int `json:"page"`
	Limit      int `json:"limit"`
	Total      int `json:"total"`
	TotalPages int `json:"totalPages"`
}

func ParsePagination(pageStr, limitStr, search string) PaginationParams {
	page, _ := strconv.Atoi(pageStr)
	limit, _ := strconv.Atoi(limitStr)
	if page < 1 {
		page = constants.DefaultPage
	}
	if limit < 1 {
		limit = constants.DefaultLimit
	}
	if limit > constants.MaxLimit {
		limit = constants.MaxLimit
	}
	return PaginationParams{
		Page:   page,
		Limit:  limit,
		Skip:   (page - 1) * limit,
		Search: search,
	}
}

func BuildPaginationMeta(page, limit, total int) PaginationMeta {
	totalPages := 0
	if limit > 0 {
		totalPages = (total + limit - 1) / limit
	}
	return PaginationMeta{Page: page, Limit: limit, Total: total, TotalPages: totalPages}
}
